rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,392 @@
+```
+
+Note that this hoisting into commands only works because `pull` and `push` are
+head/tail of the pipeline.
+
+### Mutating existing data
+
+Security telemetry data is often messy. After the data collection aspects have
+been sorted out and analysts interact with the data, it becomes clear that some
+part of the data needs to be cleaned. This matches ELT mindset of dumping
+everything first so that the data is there, and then tweak as you go.
+
+Example operations include, migrating data to a different schema, enriching
+telemetry (e.g., with GeoIP data), and fixing invalid entries.
+
+Mutation of data is non-trivial in that it requires exclusive access to the
+data. Otherwise data races may occur. Consequently, we need transactional
+interface (ACID) to support this operation.
+
+As long as VAST owns the underlying data (i.e., only VAST is allowed to make
+changes), VAST can already mutate of data at rest. [Spatial
+compaction][mutate-at-rest] uses a pre-defined disk quota as trigger to apply
+pipelines to a subset of to-be-transformed data. After VAST applies the
+pipeline, VAST optionally removes the original data in an atomic fashion.
+
+[mutate-at-rest]: https://vast.io/docs/use-vast/transform#modify-data-at-rest
+
+Modeled after the `compaction` plugin, we may consider exposing a mutable
+pipeline interface through a dedicate command. This command interacts with a
+node, and could be seen as a combination of `pull` and `push` that executes
+remotely. For example:
+
+```bash
+vast mutate 'where 10.0.0.0/8 | put orig.h resp.h'
+```
+
+This would apply the pipeline to filter and project in schema types `A`, `B`,
+and `C`.
+
+## Implementation
+
+We could implement the entire pipeline with Arrow IPC streams in the "interior"
+operators, with only `from` and `to` taking raw input as byte stream. But this
+may be too restrictive. Let's assume that input and output are dynamically
+typed. An operator must then define its input and output types, e.g., `from<In,
+Out>` takes as input `In` and produces instances of `Out`.
+
+### Case Study: Matcher
+
+The `matcher` plugin is an analyzer plugin that operator on the input stream. We
+can also rewrite the matcher use cases in term of pipelines, albeit with a
+slight change in CLI.
+
+#### Local execution
+
+Let's assume that the matcher introduces the following new pipeline operators:
+
+- `build<Arrow, Matcher>`: consumes a data stream to produce a matcher state object
+- `match<Arrow, Arrow>`: produces sightings for the configured matchers
+
+Now consider the use case of constructing a matcher from CSV data:
+
+```bash
+vast exec 'from file:///tmp/blacklist.csv |
+           read csv |
+           build --extract=net.src.ip' > matcher.flatbuf
+```
+
+After having constructed a matcher, we can now use it in a pipeline for
+matching:
+
+```bash
+vast exec 'from kafka -t /zeek/conn |
+           read zeek |
+           match --state matcher.flatbuf --on=:addr
```

Comment:
> (maybe more matcher-related than for this PR, though).

The matcher is a good case study, so we should get the pipeline-related UX right in here.

The state can be a Bloom filter, a TF model, or something else that's opaque. I see it as named reference that VAST must know about upon execution.

Can you make a suggestion how this could made clearer?

Reply:
I agree coming up with something different is difficult :-). Configuration, or ruleset, both possibly hint at "some text file in a specific form".

Could we maybe say that the "opaque state" is actually the matcher? Because IIUC, `matcher` must be able to make sense of whatever is in that `state`, be it a bloom filter or something entirely different.

So you can create or compile a matcher, and then execute it. OTOH, that doesn't really fly with `--on=:addr`.

Reply:
> Could we maybe say that the "opaque state" is actually the matcher? Because IIUC, matcher must be able to make sense of whatever is in that state, be it a bloom filter or something entirely different.

I like this terminology and adapted this.

> So you can create or compile a matcher, and then execute it. OTOH, that doesn't really fly with `--on=:addr`.

If we assume that the pipeline writer must reshape the data upstream, then the `match` operator becomes really simply and we don't need `--on`. I updated the proposal, take a look.

Reply:
> If we assume that the pipeline writer must reshape the data upstream, then the match operator becomes really simply and we don't need --on. I updated the proposal, take a look.

Looks good on first sight. Question, though: wouldn't this reshaping of the data prevent the sightings from including additional columns that are not relevant to the matcher?

Reply:
Good point, I completely forgot this. I brought back `--on` as positional arguments to `match`.

Reply:
Resolving this thread, as it looks like everything is sufficiently addressed. @dispanser please reopen if you disagree.

Reply:
perfect, thanks!